source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -396,19 +411,53 @@ void Filter::initialize(Network::ReadFilterCallbacks& callbacks, bool set_connec
   ASSERT(getStreamInfo().getDownstreamBytesMeter() == nullptr);
   ASSERT(getStreamInfo().getUpstreamBytesMeter() != nullptr);
 
-  const StreamInfo::BoolAccessor* receive_before_connect =
-      read_callbacks_->connection()
-          .streamInfo()
-          .filterState()
-          ->getDataReadOnly<StreamInfo::BoolAccessor>(ReceiveBeforeConnectKey);
-
-  // If receive_before_connect is set, we will not read disable the downstream connection
-  // as a filter before TCP_PROXY has set this state so that it can process data before
-  // the upstream connection is established.
-  if (receive_before_connect != nullptr && receive_before_connect->value()) {
-    ENVOY_CONN_LOG(debug, "receive_before_connect is enabled", read_callbacks_->connection());
+  // Initialize connection establishment mode.
+  connect_mode_ = config_->upstreamConnectMode();
+
+  // Check if early data buffering is enabled.
+  if (config_->maxEarlyDataBytes().has_value()) {
     receive_before_connect_ = true;
+    max_buffered_bytes_ = config_->maxEarlyDataBytes().value();
+    using_new_api_ = true;
+    ENVOY_CONN_LOG(debug, "receive_before_connect enabled with max buffer size: {}",
+                   read_callbacks_->connection(), max_buffered_bytes_);
   } else {
+    // Legacy behavior: check filter state for receive_before_connect.
+    const StreamInfo::BoolAccessor* receive_before_connect =
+        read_callbacks_->connection()
+            .streamInfo()
+            .filterState()
+            ->getDataReadOnly<StreamInfo::BoolAccessor>(ReceiveBeforeConnectKey);
+
+    if (receive_before_connect != nullptr && receive_before_connect->value()) {
+      ENVOY_CONN_LOG(debug, "receive_before_connect is enabled (legacy)",
+                     read_callbacks_->connection());
+      receive_before_connect_ = true;
+      using_new_api_ = false; // Legacy mode
+      // Use default buffer size for legacy mode.
+      max_buffered_bytes_ = 65536;
+    }
+  }
+
+  // Handle TLS handshake wait mode.
+  using Mode = envoy::extensions::filters::network::tcp_proxy::v3::UpstreamConnectMode;
```

Comment:
Type alias make sense. Could you create a namespace level alias like `using UpstreamConnectMode = envoy::extensions::filters::network::tcp_proxy::v3::UpstreamConnectMode;` then we can use it at the whole namespace.